packages/react-i18n/src/index.tsx

```diff
@@ -0,0 +1,68 @@
+/**
+ * External dependencies
+ */
+import * as React from 'react';
+import { __, _n, _nx, _x } from '@wordpress/i18n';
+
+export const I18nContext = React.createContext< undefined | string >( undefined );
+
+export interface I18nProps {
+	__: typeof __;
+	_n: typeof _n;
+	_nx: typeof _nx;
+	_x: typeof _x;
+	i18nLocale?: string;
+}
+
+/**
+ * React hook providing i18n translate functions
+ *
+ * @example
+ *
+ * import { useI18n } from '@automattic/react-i18n';
+ * function MyComponent() {
+ *   const { __ } = useI18n();
+ *   return <div>{ __( 'Translate me.', 'text-domain' ) }</div>;
+ * }
+ */
+export const useI18n = (): I18nProps => {
+	const locale = React.useContext( I18nContext );
+	const [ i18n, setI18n ] = React.useState< I18nProps >( makeI18n( locale ) );
+	React.useEffect( () => setI18n( makeI18n( locale ) ), [ locale ] );
```

Comment:
You shouldn't need effects and local state here at all. If the locale in context changes, the whole tree should be rerendered in one pass, triggered by the context change.

With this code, if the provider is rerendered with new value `cs` instead of the old `en`, all components that use `useI18n` will be rerendered (`useContext` will trigger that). The returned `i18n` object will still be the old one (state wasn't updated yet), its `i18nLocale` field will be `en`, but `__()` will be returning `cs` translations, because the global `@wordpress/i18n` has already loaded them!

Then the state updates triggered by `useEffect` will start to fire, individually for each component. Each of them will be rerendered again, with the updated `i18n` object, where the `i18nLocale` field is finally `cs`.

That's quite messy, isn't it? 🙂 

I recommend storing the whole `{ __, _x, ..., locale }` object in the context, not just the locale string. Then the `useI18n` hook becomes trivial:
```js
const useI18n = () => useContext( I18nContext );
```
and all the logic is in the provider. The provider will be a wrapper around the context provider, maintaining a local derived state (using `useMemo`) mapping the locale string to the `i18n` object.

The [localized-moment](https://github.com/Automattic/wp-calypso/tree/master/client/components/localized-moment) package is a good blueprint for this. The `<MomentProvider>` is much more complicated than we need here, because it also handles async loading of the locale data. When the provider gets rerendered from:
```jsx
<MomentProvider currentLocale="en" />
```
to:
```jsx
<MomentProvider currentLocale="cs" />
```
the context update and rerendering of the subtree doesn't happen immediately, but only after the new locale is loaded from the network (as a dynamic webpack chunk).

Reply:
👍 Good feeback here, thank you.

> The `<MomentProvider>` is much more complicated than we need here, because it also handles async loading of the locale data. When the provider gets rerendered from:

That's very helpful! I expect the provider to become very similar to to the `MomentProvider`. We need a way to fetch new locales and update the translated strings.

I was planning to make the provider accept an async function that resolves with updated locales, where the provider would handle updating `@wordpress/i18n` locales:

```tsx
<I18nProvider locale={ currentLocale } onLocaleChange={ fetchLocalesFromNetwork } />
```

In Calypso, `fetchLocalesFromNetwork` would contain something analagous to `switchLocale`:

https://github.com/Automattic/wp-calypso/blob/5a5fb58300b6b19848992517fffc36fc4e4b4fb9/client/lib/i18n-utils/switch-locale.js#L111-L154

But other projects could customize behavior.

Reply:
I've implemented what's outlined in 98a6dd0